File: src/propagators.py
import numpy as np
import scipy
import time
import pickle

from classes import TrajectorySH, SimulationSH
from hopping import get_hopping_probability_state_coeff, check_hop, check_hop_mash, adjust_velocity_and_hop, decoherence_zhu
from io_methods import write_to_xyz, back_up_step
from abinitio import get_electronic_structure
from kinematics import *

import fortran_modules.propagators_f as propagators_f


def velocity_verlet_position_sh(traj: TrajectorySH, ctrl: SimulationSH):
    traj.geo.position_ad += \
        traj.geo.velocity_ad*ctrl.dt + 0.5*traj.geo.force_sad[traj.hop.active]/traj.geo.mass_a[:,None]*ctrl.dt**2
    
def velocity_verlet_velocity_sh(traj: TrajectorySH, ctrl: SimulationSH):
    traj.geo.velocity_ad += \
        0.5*traj.geo.force_sad[traj.hop.active]/traj.geo.mass_a[:,None]*ctrl.dt

def update_tdc_nac_dr(traj: TrajectorySH):
    traj.est.pes.nac_ddt_ss = np.zeros((traj.est.n_states, traj.est.n_states))
    for s1 in range(traj.est.n_states):
        for s2 in range(traj.est.n_states):
            if s1 != s2:
                traj.est.pes.nac_ddt_ss[s1,s2] = np.sum(traj.geo.velocity_ad[:,:] * traj.est.pes.nac_ddr_ssad[s1,s2,:,:])/ \
                    (traj.est.pes.ham_diag_ss[s2] - traj.est.pes.ham_diag_ss[s1])

def get_state_coeff_values(traj: TrajectorySH, ctrl: SimulationSH, shift: float, prev: np.ndarray):
    temp_coeff_s = traj.hop.state_coeff_s + shift*prev*ctrl.dtq
    frac = (ctrl.qstep+shift)/ctrl.quantum_resolution
    energy_s = frac*traj.est.pes.ham_diag_ss + (1-frac)*traj.est.pes_old.ham_diag_ss
    nac_dt_ss = frac*traj.est.pes.nac_ddt_ss + (1-frac)*traj.est.pes.nac_ddt_ss
    coeff_s = np.zeros(traj.est.n_states, dtype=complex)
    for s in range(traj.est.n_states):
        coeff_s[s] = -1.j*energy_s[s,s]*temp_coeff_s[s] - np.inner(nac_dt_ss[s,:], temp_coeff_s[:])
    return coeff_s

def state_coeff_rk4(traj: TrajectorySH, ctrl: SimulationSH):
    k1 = get_state_coeff_values(traj, ctrl, 0, 0)
    k2 = get_state_coeff_values(traj, ctrl, 0.5, k1)
    k3 = get_state_coeff_values(traj, ctrl, 0.5, k2)
    k4 = get_state_coeff_values(traj, ctrl, 1, k3)
    traj.hop.state_coeff_s += ctrl.dtq/6*(k1+2*k2+2*k3+k4)

def propagate_state_coeff(traj: TrajectorySH, ctrl: SimulationSH):
    for ctrl.qstep in range(ctrl.quantum_resolution):
        #propagators_f.state_coeff_rk4(traj.hop.state_coeff_s, traj.est.pes.nac_dt_ss, traj.est.pes.energy_s, ctrl.dtq)
        state_coeff_rk4(traj, ctrl)
        if traj.hop.target == traj.hop.active:
            get_hopping_probability_state_coeff(traj, ctrl.dtq)
            check_hop(traj)
    #propagators_f.normalise_state_coeff(traj.hop.state_coeff_s)

def initialise_dynamics(traj: TrajectorySH, ctrl: SimulationSH):
    traj.est.pes_old = traj.est.pes
    get_electronic_structure(traj, ctrl)
    traj.est.first = False

    back_up_step(traj, ctrl)
    write_to_xyz(traj, ctrl)

def loop_dynamics(traj: TrajectorySH, ctrl: SimulationSH):
    while not (lambda ctrl, traj: eval(ctrl.termination_cond))(ctrl, traj):
        print(traj.est.pes.ham_diag_ss)
        print(traj.est.pes.nac_ddr_ssad)

        t1 = time.time()
        # step 1: position update & 1/2 velocity update
        propagators_f.update_position_velocity_verlet(
            traj.geo.position_ad, traj.geo.velocity_ad, traj.geo.force_sad[traj.hop.active], traj.geo.mass_a, ctrl.dt
        )
        propagators_f.update_velocity_velocity_verlet(
            traj.geo.velocity_ad, traj.geo.force_sad[traj.hop.active], traj.geo.mass_a, ctrl.dt
        )

        t2 = time.time()
        # step 2: electronic structure
        traj.est.pes_old = traj.est.pes
        get_electronic_structure(traj, ctrl)

        t3 = time.time()
        # step 3: 2/2 velocity update
        propagators_f.update_velocity_velocity_verlet(
            traj.geo.velocity_ad, traj.geo.force_sad[traj.hop.active], traj.geo.mass_a, ctrl.dt
        )

        t4 = time.time()
        # step 4&5: update electronic wf coefficients & compute hopping probabilities
        traj.est.pes.nac_ddt_ss = propagators_f.update_tdc(traj.geo.velocity_ad, traj.est.pes.nac_ddr_ssad)
        propagate_state_coeff(traj, ctrl)

        t5 = time.time()
        # step 6: select a new active state
        adjust_velocity_and_hop(traj)
        decoherence_zhu(traj, ctrl.dt)

        t6 = time.time()

        # step 7: recalculate EST if hop occured
        if traj.est.recalculate:
            get_electronic_structure(traj, ctrl)
            traj.est.recalculate = False

        # increment time
        ctrl.current_time += ctrl.dt
        ctrl.step += 1
        back_up_step(traj, ctrl)
        write_to_xyz(traj, ctrl)

        if 0:
            print("nuclear 1/2: ", t2-t1)
            print("est: ", t3-t2)
            print("nuclear 2/2: ", t4-t3)
            print("coeff: ", t5-t4)
            print("active: ", t6-t5)
            print("total: ", t6-t1, "\n")

File: fmodules/propagators.f90
subroutine update_position_velocity_verlet(position_ad, velocity_ad, force_ad, mass_a, dt, n_atoms)
    implicit none
    
    !f2py intent(in) n_atoms
    integer, intent(in) :: n_atoms
    !f2py intent(inout) position_ad
    real*8, dimension(n_atoms, 3), intent(inout) :: position_ad
    !f2py intent(in) velocity_ad
    real*8, dimension(n_atoms, 3), intent(in) :: velocity_ad
    !f2py intent(in) force_ad
    real*8, dimension(n_atoms, 3), intent(in) :: force_ad
    !f2py intent(in) mass_a
    real*8, dimension(n_atoms), intent(in) :: mass_a
    !f2py intent(in) dt
    real*8, intent(in) :: dt

    integer a, d
    do a = 1, n_atoms
        do d = 1, 3
            position_ad(a,d) = position_ad(a,d) + velocity_ad(a,d)*dt + 0.5*force_ad(a,d)/mass_a(a)*dt**2
        end do
    end do
end subroutine update_position_velocity_verlet

subroutine update_velocity_velocity_verlet(velocity_ad, force_ad, mass_a, dt, n_atoms)
    implicit none
    
    !f2py intent(in) n_atoms
    integer, intent(in) :: n_atoms
    !f2py intent(inout) velocity_ad
    real*8, dimension(n_atoms, 3), intent(inout) :: velocity_ad
    !f2py intent(in) force_ad
    real*8, dimension(n_atoms, 3), intent(in) :: force_ad
    !f2py intent(in) mass_a
    real*8, dimension(n_atoms), intent(in) :: mass_a
    !f2py intent(in) dt
    real*8, intent(in) :: dt

    integer a, d
    do a = 1, n_atoms
        do d = 1, 3
            velocity_ad(a,d) = velocity_ad(a,d) + 0.5*force_ad(a,d)/mass_a(a)*dt
        end do
    end do
end subroutine update_velocity_velocity_verlet

subroutine get_state_coeff_values(coeff_in_s, coeff_out_s, nac_dt_ss, energy_s, n_states)
    implicit none
    
    !f2py intent(in) n_states
    integer, intent(in) :: n_states
    !f2py intent(in) nac_dt_ss
    real*8, dimension(n_states, n_states), intent(in) :: nac_dt_ss
    !f2py intent(in) energy_s
    real*8, dimension(n_states), intent(in) :: energy_s
    !f2py intent(in) coeff_in_s
    complex*16, dimension(n_states), intent(in) :: coeff_in_s
    !f2py intent(out) coeff_out_s
    complex*16, dimension(n_states), intent(out) :: coeff_out_s

    integer s
    complex*16 :: ii = (0,1)
    do s = 1, n_states
        coeff_out_s(s) = -ii*energy_s(s)*coeff_in_s(s) - sum(nac_dt_ss(s,:)*coeff_in_s(:))
    end do
end subroutine 

subroutine state_coeff_rk4(state_coeff_s, nac_dt_ss, energy_s, dtq, n_states)
    implicit none
    
    !f2py intent(in) n_states
    integer, intent(in) :: n_states
    !f2py intent(in) nac_dt_ss
    real*8, dimension(n_states, n_states), intent(in) :: nac_dt_ss
    !f2py intent(in) energy_s
    real*8, dimension(n_states), intent(in) :: energy_s
    !f2py intent(inout) coeff_in_s
    complex*16, dimension(n_states), intent(inout) :: state_coeff_s
    !f2py intent(in) dtq
    real*8, intent(in) :: dtq

    complex*16, dimension(n_states) :: k1, k2, k3, k4, temp
    call get_state_coeff_values(state_coeff_s, k1, nac_dt_ss, energy_s, n_states)
    temp = state_coeff_s + dtq*k1/2
    call get_state_coeff_values(temp, k2, nac_dt_ss, energy_s, n_states)
    temp = state_coeff_s + dtq*k2/2
    call get_state_coeff_values(temp, k3, nac_dt_ss, energy_s, n_states)
    temp = state_coeff_s + dtq*k3
    call get_state_coeff_values(temp, k4, nac_dt_ss, energy_s, n_states)
    state_coeff_s = state_coeff_s + dtq/6*(k1 + 2*k2 + 2*k3 + k4)
end subroutine state_coeff_rk4

subroutine update_tdc(velocity_ad, nac_dr_ssad, nac_dt_ss, n_states, n_atoms)
    !f2py intent(in) n_states
    integer, intent(in) :: n_states
    !f2py intent(in) n_atoms
    integer, intent(in) :: n_atoms
    !f2py intent(in) velocity_ad
    real*8, intent(in), dimension(n_atoms, 3) :: velocity_ad
    !f2py intent(in) nac_dr_ssad
    real*8, intent(in), dimension(n_states, n_states, n_atoms, 3) :: nac_dr_ssad
    !f2py intent(out) nac_dt_ss
    real*8, intent(out), dimension(n_states, n_states) :: nac_dt_ss

    integer s1, s2
    do s1 = 1, n_states
        do s2 = 1, n_states
            if (s1 .ne. s2) then
                nac_dt_ss(s1,s2) = sum(velocity_ad(:,:) * nac_dr_ssad(s1,s2,:,:))
            else
                nac_dt_ss(s1,s2) = 0.
            end if
        end do
    end do
end subroutine update_tdc

subroutine update_overlap(state_old_ss, state_new_ss, overlap_ss, n_states)
    !f2py intent(in) n_states
    integer, intent(in) :: n_states
    !f2py intent(in) state_old_ss
    real*8, intent(in), dimension(n_states, n_states) :: state_old_ss
    !f2py intent(in) state_new_ss
    real*8, intent(in), dimension(n_states, n_states) :: state_new_ss
    !f2py intent(out) overlap_ss
    real*8, intent(out), dimension(n_states, n_states) :: overlap_ss
    
    integer s1, s2
    do s1 = 1, n_states
        do s2 = 1, n_states
            overlap_ss(s1, s2) = sum(state_old_ss(:,s1) * state_new_ss(:,s2))
        end do
    end do
end subroutine update_overlap

subroutine normalise_state_coeff(state_coeff_s, n_states)
    !f2py intent(in) n_states
    integer, intent(in) :: n_states
    !f2py intent(inout) state_coeff_s
    complex*16, intent(inout), dimension(n_states) :: state_coeff_s

    real*8 norm

    norm = sum(abs(state_coeff_s)**2)
    state_coeff_s = state_coeff_s/norm
end subroutine normalise_state_coeff